Add NGINX reverse proxy

surminus opened the pull request:
I tried in vain to use Traefik as a small reverse proxy since it forwarded error pages out of the box, and is built for cloud environments.

Unfortunately I ran into issues with the `Authorization` header being set when requesting the error page, and S3 failing since that header was set.

This replaces the functionality with Good Ole NGINX, and strips the `Authorization` header when we're passing on a request to an error page.

I've defined every single error page in a separate block with a separate `upstream` block for the static pages, since I ran into issues trying to be clever, where I had to set `resolver` configuration.

Basically, I got bored and decided not to be clever.

surminus committed:
Add NGINX reverse proxy

I tried in vain to use Traefik as a small reverse proxy since it
forwarded error pages out of the box, and is built for cloud
environments.

Unfortunately I ran into issues with the `Authorization` header being
set when requesting the error page, and S3 failing since that header was
set.

This replaces the functionality with Good Ole NGINX, and strips the
`Authorization` header when we're passing on a request to an error page.

I've defined every single error page in a separate block with a separate
`upstream` block for the static pages, since I ran into issues trying to
be clever, where I had to set `resolver` configuration.

Basically, I got bored and decided not to be clever.

surminus commented:
One thing I've realised is we may need to set real ip configuration? Lots of stuff I was hoping to remove has sadly come back to haunt us with the introduction of NGINX.

elenatanasoiu commented:
Followed by:
- https://github.com/futurelearn/deployment/pull/81
- https://github.com/futurelearn/terraform/pull/257
Related to:
- https://github.com/futurelearn/architecture-decision-record/pull/4

elenatanasoiu commented on nginx-proxy/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM nginx
+
+RUN apt-get update && apt-get install -y netcat
```
What are we doing with `netcat`?

elenatanasoiu replied:
Ah https://github.com/futurelearn/deployment/pull/81/commits/f5a19f350b0f541dd0b92387980d29d0b66ff9a3#diff-fcc9f67616f16fa4f6a6196995e4c3ccR100

surminus replied:
Yeah, healthchecks.

elenatanasoiu commented on nginx-proxy/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+ENTRYPOINT ["/opt/entrypoint.sh"]
+
+CMD ["nginx", "-g", "daemon off;"]
```
Will this be changed to `daemon on` at some point? Or is the idea to have this container just running `nginx` so it should be in the foreground all the time?

surminus replied:
It should run as a foreground process in a container. The container should basically run a single process, so there is no need for it to run in the background.

elenatanasoiu commented on nginx-proxy/README.md:
```diff
@@ -0,0 +1,3 @@
+# nginx-proxy
+
+A small reverse to forward error pages to an external location.
```
reverse-proxy* ?

elenatanasoiu commented on nginx-proxy/default.conf:
```diff
@@ -0,0 +1,103 @@
+  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+  proxy_set_header X-Real-IP  $remote_addr;
+  proxy_set_header Host $http_host;
+  proxy_set_header If-Modified-Since $http_if_modified_since;
```
I understand we need to overwrite the `Authorization` header and `Host` header for each error page. Why are we setting these particular 4 headers as well?

surminus replied:
I copied this from our [current NGINX config](https://github.com/futurelearn/chef/blob/master/site-cookbooks/app/templates/default/futurelearn_nginx_conf.erb#L141-L144).

The main thing is `X-Forwarded-For` - I wanted to ensure we continued forwarding the remote user IP.

elenatanasoiu commented on nginx-proxy/entrypoint.sh:
```diff
@@ -0,0 +1,15 @@
+[[ -z "${BACKEND_HOST}" ]] && export BACKEND_HOST="127.0.0.1"
+
+# shellcheck disable=SC2016
+envsubst '${BACKEND_HOST} ${STATIC_PAGES_HOST} ${PORT}' < /etc/nginx/default.conf.template > /etc/nginx/conf.d/default.conf
```
Nice! TIL

surminus committed:
Set real IP configuration

We want to ensure the users remote IP is forwarded rather than an IP
from the load balancer or Cloudflare.

This users the set_real_ip_from config option[1] to ensure this happens.

We populate the configuration with Cloudflare addresses, and the private
network range used by our load balancers internally.

[1] https://nginx.org/en/docs/http/ngx_http_realip_module.html

surminus commented:
I added a commit for real ip configuration
